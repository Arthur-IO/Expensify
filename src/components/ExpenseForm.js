import React from "react";
import moment from 'moment'
import {SingleDatePicker} from 'react-dates';

// A form that takes in data to create a new expense

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: false
        }
    }

    /**
     * An action that is fired when the description input is changed
     * @param e The [Event] that is passed into this function
     */
    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({description}))
    }
    /**
     * An action that is fired when the note input is changed
     * @param e The [Event] that is passed into this function
     */
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({note}))
    }
    /**
     * An action that is fired when the amount input is changed
     * @param e The [Event] that is passed into this function
     */
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    }
    /**
     * An action that is fired when the date selector is changed
     * @param createdAt     The date selected by the date picker
     */
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt}))
        }
    }
    /**
     * An action that is fired when the calendar input is clicked
     * @param calendarFocused       A flag that determines if the date picker is focused
     */
    onFocusChange = ({focused: calendarFocused}) => {
        this.setState(() => ({calendarFocused}))
    }

    /**
     * Submits this form to attempt to add it to the expense list
     * @param e The [Event] that is passed into this function
     */
    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            // Displays an error if the following inputs have no value
            this.setState(() => ({error: true}))
        } else {
            // Clears a potential error if the following inputs have no value
            this.setState(() => ({error: false}))

            // Submits this expense with the provided information
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    render() {
        return (
            <form
                className={"form"}
                onSubmit={this.onSubmit}
            >
                {
                    this.state.error && <p className={"form__error"}>Please provide description and amount</p>
                }
                <input
                    className={"text-input"}
                    type={"text"}
                    placeholder={"Description"}
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    className={"text-input"}
                    type={"text"}
                    placeholder={"Amount"}
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className={"textarea"}
                    placeholder={"Add a note for your expense (optional)"}
                    value={this.state.note}
                    onChange={this.onNoteChange}
                />
                <div>
                    <button className={"button"}>Save Expense</button>
                </div>
            </form>
        )
    }
}