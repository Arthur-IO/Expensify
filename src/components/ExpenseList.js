import React from "react";
import {connect} from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses'

// A list of expense items

export const ExpenseList = (props) => (
    <div className={"content-container"}>
        <div className={"list-header"}>
            <div className={"show-for-mobile"}>Expenses</div>
            <div className={"show-for-desktop"}>Expense</div>
            <div className={"show-for-desktop"}>Amount</div>
        </div>
        <div className={"list-body"}>
            {
                // Display 'no expenses' message when there are no expenses
                props.expenses.length === 0 ? (
                    <div className={"list-item list-item--message"}>
                        <span>No expenses</span>
                    </div>
                ) : (
                    // The individual expense list items rendered
                    props.expenses.map(expense => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
            }
        </div>
    </div>
)

// Maps redux state to the props of this component
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

// Connects this component to the redux store and passes in redux state to the component props
export default connect(mapStateToProps)(ExpenseList)