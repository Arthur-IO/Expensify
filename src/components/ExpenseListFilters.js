import React from "react";
import {connect} from "react-redux";
import {setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate} from "../actions/filters";
import {DateRangePicker} from 'react-dates'

// The ExpenseList filters to filter or sort list items

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChanged = (focusedInput) => {
        this.setState({
            calendarFocused: focusedInput
        })
    }

    onTextChanged = (e) => {
        const filterText = e.target.value
        this.props.setTextFilter(filterText)
    }

    onSortByChanged = (e) => {
        const value = e.target.value
        if (value === 'date') {
            this.props.sortByDate()
        } else {
            this.props.sortByAmount()
        }
    }

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChanged}/>
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChanged}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChanged}
                    isOutsideRange={() => false}
                    numberOfMonths={1}
                    showClearDates={true}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (filterText) => dispatch(setTextFilter(filterText)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)