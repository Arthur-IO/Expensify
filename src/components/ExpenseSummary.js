import React from 'react'
import numeral from 'numeral'
import {connect} from 'react-redux'
import selectExpenses from '../selectors/expenses'
import getTotalExpenses from "../selectors/expenses-total";

export const ExpenseSummary = (props) => {
    const totalViewing = props.totalExpenses ? props.totalExpenses : 0
    const expensesWord = totalViewing === 1 ? 'expense' : 'expenses'
    const formattedPrice = numeral(props.expensesTotal / 100).format('$0,0.00')

    return (
        <div>
            <p>Viewing {totalViewing} {expensesWord} with a total of {formattedPrice}
            </p>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)

    return {
        totalExpenses: visibleExpenses.length,
        expensesTotal: getTotalExpenses(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)