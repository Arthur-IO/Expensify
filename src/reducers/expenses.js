// Expenses reducer - Dispatches actions to the redux store for provided and supported arguments

const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // Add item using spread operators
            return [
                ...state,
                action.expense
            ]
        case 'REMOVE_EXPENSE':
            return state.filter(expense => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense
                }
            })
        default:
            return state
    }
}