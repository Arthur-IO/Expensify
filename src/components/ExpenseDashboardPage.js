import React from 'react';
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from "./ExpenseSummary";

// The home-page for this application

export const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
);

export default ExpenseDashboardPage;
