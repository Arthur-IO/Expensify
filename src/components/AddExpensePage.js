import React from 'react';
import {connect} from 'react-redux'

import ExpenseForm from "./ExpenseForm";
import {addExpense} from "../actions/expenses";

// Adds an expense to the expenses list

export class AddExpensePage extends React.Component {

    onSubmit = (expense) => {
        // Dispatches an event to the redux store
        this.props.addExpense(expense)

        // Redirects the user back to the home-page
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }

}

// Maps dispatch to the properties of this component
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

// Connects this component to the redux store with no props
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
