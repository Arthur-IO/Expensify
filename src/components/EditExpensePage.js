import React from 'react';
import {connect} from 'react-redux'
import ExpenseForm from "./ExpenseForm";
import {editExpense, removeExpense} from "../actions/expenses";

// A page where the user can edit a specific expense in their expense list

export class EditExpensePage extends React.Component {

    constructor(props) {
        super(props);
    }

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push("/")
    }

    onClick = () => {
        this.props.removeExpense({ id: this.props.expense.id })
        this.props.history.push("/")
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>
                    Remove
                </button>
            </div>
        )
    }

}

// Maps the redux state to the props of the individual component
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(expense => expense.id === props.match.params.id)
    }
}

// Maps the dispatch function to the properties of this component
const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, updates) => dispatch(editExpense(id, updates)),
    removeExpense: ({ id }) => dispatch(removeExpense({ id }))
})

// Connects this component to the redux store with the redux state passed in as props
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
