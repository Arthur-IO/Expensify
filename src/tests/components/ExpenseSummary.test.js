import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseSummary} from "../../components/ExpenseSummary";
import expenses from "../fixtures/expenses";
import getExpensesTotal from '../../selectors/expenses-total'

test('should render ExpenseSummary correctly with no expenses', () => {
    const wrapper = shallow(
        <ExpenseSummary
            totalExpenses={[]}
            expensesTotal={0}
        />
    )
    expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseSummary correctly with expenses', () => {
    const expensesTotal = getExpensesTotal(expenses)
    const wrapper = shallow(
        <ExpenseSummary
            totalExpenses={expenses}
            expensesTotal={expensesTotal}
        />
    )
    expect(wrapper).toMatchSnapshot()
})