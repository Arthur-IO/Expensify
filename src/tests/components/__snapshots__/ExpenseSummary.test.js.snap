// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with expenses 1`] = `
<div>
  <p>
    Viewing 
     
    expenses
     with a total of 
    $11.50
  </p>
</div>
`;

exports[`should render ExpenseSummary correctly with no expenses 1`] = `
<div>
  <p>
    Viewing 
     
    expenses
     with a total of 
    $0.00
  </p>
</div>
`;
